### PR TITLE
IR: refactor InitializersLoweringBase and descendants

### PR DESCRIPTION
**Pull request by pyos**
Also, in JVM_IR specifically do not add `const val` initializers to `<clinit>`, since they are implicitly generated by the JVM from ConstantValue attributes.

### CODE REVIEW

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt [hunk 1]
```diff
@@ -7,84 +7,72 @@ package org.jetbrains.kotlin.backend.common.lower
...
     }
 }
 
-abstract class InitializersLoweringBase(val context: CommonBackendContext) : ClassLoweringPass {
-    protected fun handleDeclaration(irClass: IrClass, declaration: IrDeclaration): IrStatement? = when (declaration) {
-        is IrField -> handleField(irClass, declaration)
-        is IrAnonymousInitializer -> handleAnonymousInitializer(declaration)
-        else -> null
+abstract class InitializersLoweringBase(open val context: CommonBackendContext) : ClassLoweringPass {
+    protected fun extractInitializers(irClass: IrClass, filter: (IrDeclaration) -> Boolean): List<IrStatement> {
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt]: I'd prefer to make it more readable, something like this, what do you think?
```
    protected fun extractInitializers(irClass: IrClass, filter: (IrDeclaration) -> Boolean): List<IrStatement> =
        irClass.declarations.filter(filter).mapNotNull {
            when (it) {
                is IrField -> handleField(irClass, it)
                is IrAnonymousInitializer -> handleAnonymousInitializer(it)
                else -> null
            }
        }.also { 
            irClass.declarations.removeAll { it is IrAnonymousInitializer }
        }
```

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt [hunk 2]
```diff
@@ -7,84 +7,72 @@ package org.jetbrains.kotlin.backend.common.lower
...
 
-    private fun transformInstanceInitializerCallsInConstructors(irClass: IrClass, instanceInitializerStatements: List<IrStatement>) {
-        irClass.transformChildrenVoid(object : IrElementTransformerVoid() {
-            override fun visitInstanceInitializerCall(expression: IrInstanceInitializerCall): IrExpression {
-                val copiedBlock =
-                    IrBlockImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, context.irBuiltIns.unitType, null, instanceInitializerStatements)
-                        .copy(irClass) as IrBlock
-                return IrBlockImpl(irClass.startOffset, irClass.endOffset, context.irBuiltIns.unitType, null, copiedBlock.statements)
-            }
+            override fun visitSimpleFunction(declaration: IrSimpleFunction) = declaration
```
Comment (udalov) [inline compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt]: Could you explain why this line is needed?

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt (hunk 1 above)
Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt]: Sure. My intention was to not have any list copies, but I guess that's my C++ background making itself visible. (Though note that your snippet is slightly incorrect - `removeAll` should also check `filter(it)`.)

File: compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt (hunk 2 above)
Comment (pyos) [reply compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt]: Same reason as `visitClassNew`: to not visit the parts of the tree that don't have any instance initializer calls anyway. The comment above applies to both of them, which would be clearer if there was no empty line, but then IntelliJ would complain about a style guide violation...

**Comment by pyos** [comment]
@udalov I've also added a change in ConstLowering to this PR (the last commit) which is not relevant to InitializersLowering, but is just 3 lines and would otherwise have a merge conflict.

**Comment by udalov** [comment]
@pyos Looks good, merged manually, thanks!

**Pull request closed without merging**